src/nvm.rs

```diff
@@ -269,31 +288,35 @@ impl<T, const N: usize> Collection<T, N> where T: Copy {
         }
     }
 
-    /// Returns true if the indicated slot is allocated, or false if it is
-    /// free.
-    pub fn is_allocated(&self, index: usize) -> bool {
-        self.flags.get_ref()[index] == STORAGE_VALID
+    /// Returns `Ok(())` if the indicated slot is allocated.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error describing whether the key was out of bounds or
+    /// the indicated slot was not allocated.
+    fn is_allocated(&self, key: usize) -> Result<(), KeyError> {
+        match self.flags.get_ref().get(key) {
+            Some(&byte) => {
+                if byte == STORAGE_VALID {
+                    Ok(())
+                } else {
+                    Err(SlotIsFree)
+                }
+            },
+            None => Err(KeyOutOfRange)
+        }
     }
 
     /// Returns the number of allocated slots.
     pub fn len(&self) -> usize {
-        let mut result = 0;
-        for v in self.flags.get_ref() {
-            if *v == STORAGE_VALID {
-                result += 1;
-            }
-        }
-        result
+        self.flags.get_ref()
+            .iter()
+            .fold(0, |acc, &byte| acc + (byte == STORAGE_VALID) as u32) as usize
```

Comment:
I initially had written
```rust
self.flags.get_ref()
    .iter()
    .filter(|&byte| *byte == STORAGE_VALID)
    .count()
```
but after digging on the interwebs it turns out [this weird casting is more performant](https://docs.rs/bytecount/0.6.2/src/bytecount/naive.rs.html#11-13)